File: lib/src/groups/group_session.dart
/// Group session for Signal Protocol group messaging.
library;

import 'dart:ffi';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import '../bindings/libsignal_bindings.dart';
import '../exception.dart';
import '../ffi_helpers.dart';
import '../libsignal.dart';
import '../protocol/protocol_address.dart';
import '../stores/sender_key_store.dart';
import '../utils.dart';
import 'sender_key_distribution_message.dart';

/// A map to store sender key record bytes during FFI callback execution.
///
/// This is used to pass data between Dart and native code during
/// group encryption/decryption operations.
final _pendingSenderKeyRecordBytes = <int, Uint8List>{};
final _pendingStoreOperations = <int, _StoreOperation>{};
int _operationCounter = 0;

/// Gets a unique operation ID with overflow protection.
int _nextOperationId() {
  _operationCounter++;
  // Reset if approaching max int to prevent overflow
  if (_operationCounter > 0x7FFFFFFFFFFFFFF) {
    _operationCounter = 1;
  }
  return _operationCounter;
}

class _StoreOperation {
  final SenderKeyStore store;
  final ProtocolAddress sender;
  final Uint8List distributionId;

  _StoreOperation(this.store, this.sender, this.distributionId);
}

/// Native callback for loading sender keys.
int _loadSenderKeyCallback(
  Pointer<Void> ctx,
  Pointer<SignalMutPointerSenderKeyRecord> out,
  SignalConstPointerProtocolAddress sender,
  Pointer<Pointer<Uint8>> distributionId,
) {
  try {
    final operationId = ctx.address;
    final operation = _pendingStoreOperations[operationId];

    if (operation == null) {
      return -1; // Error
    }

    // This is synchronous - we pre-loaded the key before starting
    final recordBytes = _pendingSenderKeyRecordBytes[operationId];
    if (recordBytes != null) {
      // Deserialize the record directly via FFI without creating a wrapper
      // (to avoid double-free issues with finalizer)
      final dataPtr = calloc<Uint8>(recordBytes.length);
      dataPtr.asTypedList(recordBytes.length).setAll(0, recordBytes);

      final buffer = calloc<SignalBorrowedBuffer>();
      buffer.ref.base = dataPtr.cast<UnsignedChar>();
      buffer.ref.length = recordBytes.length;

      final recordPtr = calloc<SignalMutPointerSenderKeyRecord>();

      final error =
          signal_sender_key_record_deserialize(recordPtr, buffer.ref);

      // Free the temporary buffers (libsignal makes a copy of the data)
      calloc.free(dataPtr);
      calloc.free(buffer);

      if (error == nullptr && recordPtr.ref.raw != nullptr) {
        out.ref.raw = recordPtr.ref.raw;
        calloc.free(recordPtr); // Free the pointer holder, not the record
        return 0; // Success
      }

      calloc.free(recordPtr);
    }

    out.ref.raw = nullptr;
    return 0; // No record (not an error)
  } catch (e) {
    return -1; // Error
  }
}

/// Native callback for storing sender keys.
int _storeSenderKeyCallback(
  Pointer<Void> ctx,
  SignalConstPointerProtocolAddress sender,
  Pointer<Pointer<Uint8>> distributionId,
  SignalConstPointerSenderKeyRecord record,
) {
  try {
    final operationId = ctx.address;
    final operation = _pendingStoreOperations[operationId];

    if (operation == null) {
      return -1;
    }

    // Serialize the record to bytes because the original pointer is owned
    // by libsignal and may be freed after callback returns
    if (record.raw != nullptr) {
      final outPtr = calloc<SignalOwnedBuffer>();

      try {
        final error = signal_sender_key_record_serialize(outPtr, record);
        if (error == nullptr) {
          _pendingSenderKeyRecordBytes[operationId] =
              FfiHelpers.fromOwnedBuffer(outPtr.ref);
        }
      } finally {
        calloc.free(outPtr);
      }
    }

    return 0;
  } catch (e) {
    return -1;
  }
}

/// Converts a 16-byte UUID to a string.
String _uuidToString(Uint8List uuid) {
  if (uuid.length != 16) {
    throw ArgumentError('UUID must be 16 bytes');
  }

  final hex = uuid.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-'
      '${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20)}';
}

/// Converts a UUID string to 16 bytes.
Uint8List _stringToUuid(String uuid) {
  final hex = uuid.replaceAll('-', '');
  if (hex.length != 32) {
    throw ArgumentError('Invalid UUID string: $uuid');
  }

  final result = Uint8List(16);
  for (var i = 0; i < 16; i++) {
    result[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return result;
}

/// A group session for encrypting and decrypting group messages.
///
/// Group messaging uses sender keys to enable efficient encryption for
/// multiple recipients. Each group member maintains sender keys for all
/// other members.
///
/// Example:
/// ```dart
/// // Create a group session
/// final session = GroupSession(
///   senderAddress,
///   distributionId,
///   senderKeyStore,
/// );
///
/// // Create and distribute the sender key
/// final distMessage = await session.createDistributionMessage();
/// // Send distMessage to all group members...
///
/// // Encrypt a message
/// final encrypted = await session.encrypt(Uint8List.fromList(utf8.encode('Hello')));
/// ```
class GroupSession {
  final ProtocolAddress _senderAddress;
  final Uint8List _distributionId;
  final SenderKeyStore _store;

  /// Creates a new group session.
  ///
  /// The [senderAddress] is the address of the message sender (typically us).
  /// The [distributionId] is a UUID identifying this group/distribution.
  /// The [store] provides persistent storage for sender keys.
  GroupSession(this._senderAddress, this._distributionId, this._store) {
    LibSignal.ensureInitialized();

    if (_distributionId.length != 16) {
      throw ArgumentError('distributionId must be a 16-byte UUID');
    }
  }

  /// Creates a sender key distribution message for this group.
  ///
  /// This message should be sent to all group members so they can
  /// decrypt messages from this sender.
  Future<SenderKeyDistributionMessage> createDistributionMessage() async {
    final operationId = _nextOperationId();

    // Pre-load existing sender key record if any
    final senderKeyName = SenderKeyName(
      _senderAddress,
      _uuidToString(_distributionId),
    );
    final existingData = await _store.loadSenderKey(senderKeyName);
    if (existingData != null) {
      _pendingSenderKeyRecordBytes[operationId] = existingData;
    }

    _pendingStoreOperations[operationId] = _StoreOperation(
      _store,
      _senderAddress,
      _distributionId,
    );

    try {
      final outPtr = calloc<SignalMutPointerSenderKeyDistributionMessage>();
      final senderPtr = calloc<SignalConstPointerProtocolAddress>();
      senderPtr.ref.raw = _senderAddress.pointer;

      // The C function expects uint8_t (*distribution_id)[16] - a pointer to a
      // 16-byte array. The binding incorrectly uses Pointer<Pointer<Uint8>>,
      // so we allocate a 16-byte buffer and cast it appropriately.
      final distIdData = calloc<Uint8>(16);
      distIdData.asTypedList(16).setAll(0, _distributionId);

      // Create the store struct with callbacks
      final storePtr = calloc<SignalSenderKeyStore>();
      storePtr.ref.ctx = Pointer.fromAddress(operationId);
      storePtr.ref.load_sender_key =
          Pointer.fromFunction<SignalLoadSenderKeyFunction>(
        _loadSenderKeyCallback,
        -1,
      );
      storePtr.ref.store_sender_key =
          Pointer.fromFunction<SignalStoreSenderKeyFunction>(
        _storeSenderKeyCallback,
        -1,
      );

      final storeConstPtr = calloc<SignalConstPointerFfiSenderKeyStoreStruct>();
      storeConstPtr.ref.raw = storePtr;

      try {
        final error = signal_sender_key_distribution_message_create(
          outPtr,
          senderPtr.ref,
          distIdData.cast<Pointer<Uint8>>(),
          storeConstPtr.ref,
        );
        FfiHelpers.checkError(
          error,
          'signal_sender_key_distribution_message_create',
        );

        if (outPtr.ref.raw == nullptr) {
          throw LibSignalException.nullPointer(
            'signal_sender_key_distribution_message_create',
          );
        }

        // Save any updated sender key record
        final updatedRecordBytes = _pendingSenderKeyRecordBytes[operationId];
        if (updatedRecordBytes != null) {
          await _store.storeSenderKey(senderKeyName, updatedRecordBytes);
        }

        return SenderKeyDistributionMessage.fromPointer(outPtr.ref.raw);
      } finally {
        calloc.free(outPtr);
        calloc.free(senderPtr);
        calloc.free(distIdData);
        calloc.free(storePtr);
        calloc.free(storeConstPtr);
      }
    } finally {
      _pendingStoreOperations.remove(operationId);
      _pendingSenderKeyRecordBytes.remove(operationId);
    }
  }

  /// Processes a sender key distribution message from another group member.
  ///
  /// The [senderAddress] is the address of the member who sent this
  /// distribution message. This stores their key material so their
  /// messages can be decrypted.
  Future<void> processDistributionMessage(
    ProtocolAddress senderAddress,
    SenderKeyDistributionMessage message,
  ) async {
    final operationId = _nextOperationId();

    // Pre-load existing sender key record if any
    final senderKeyName = SenderKeyName(
      senderAddress,
      _uuidToString(_distributionId),
    );
    final existingData = await _store.loadSenderKey(senderKeyName);
    if (existingData != null) {
      _pendingSenderKeyRecordBytes[operationId] = existingData;
    }

    _pendingStoreOperations[operationId] = _StoreOperation(
      _store,
      senderAddress,
      _distributionId,
    );

    try {
      final senderPtr = calloc<SignalConstPointerProtocolAddress>();
      senderPtr.ref.raw = senderAddress.pointer;

      final msgPtr = calloc<SignalConstPointerSenderKeyDistributionMessage>();
      msgPtr.ref.raw = message.pointer;

      // Create the store struct with callbacks
      final storePtr = calloc<SignalSenderKeyStore>();
      storePtr.ref.ctx = Pointer.fromAddress(operationId);
      storePtr.ref.load_sender_key =
          Pointer.fromFunction<SignalLoadSenderKeyFunction>(
        _loadSenderKeyCallback,
        -1,
      );
      storePtr.ref.store_sender_key =
          Pointer.fromFunction<SignalStoreSenderKeyFunction>(
        _storeSenderKeyCallback,
        -1,
      );

      final storeConstPtr = calloc<SignalConstPointerFfiSenderKeyStoreStruct>();
      storeConstPtr.ref.raw = storePtr;

      try {
        final error = signal_process_sender_key_distribution_message(
          senderPtr.ref,
          msgPtr.ref,
          storeConstPtr.ref,
        );
        FfiHelpers.checkError(
          error,
          'signal_process_sender_key_distribution_message',
        );

        // Save any updated sender key record
        final updatedRecordBytes = _pendingSenderKeyRecordBytes[operationId];
        if (updatedRecordBytes != null) {
          await _store.storeSenderKey(senderKeyName, updatedRecordBytes);
        }
      } finally {
        calloc.free(senderPtr);
        calloc.free(msgPtr);
        calloc.free(storePtr);
        calloc.free(storeConstPtr);
      }
    } finally {
      _pendingStoreOperations.remove(operationId);
      _pendingSenderKeyRecordBytes.remove(operationId);
    }
  }

  /// Encrypts a message for the group.
  ///
  /// Returns the encrypted message bytes.
  Future<Uint8List> encrypt(Uint8List plaintext) async {
    final operationId = _nextOperationId();

    // Pre-load existing sender key record
    final senderKeyName = SenderKeyName(
      _senderAddress,
      _uuidToString(_distributionId),
    );
    final existingData = await _store.loadSenderKey(senderKeyName);
    if (existingData != null) {
      _pendingSenderKeyRecordBytes[operationId] = existingData;
    }

    _pendingStoreOperations[operationId] = _StoreOperation(
      _store,
      _senderAddress,
      _distributionId,
    );

    try {
      final outPtr = calloc<SignalMutPointerCiphertextMessage>();
      final senderPtr = calloc<SignalConstPointerProtocolAddress>();
      senderPtr.ref.raw = _senderAddress.pointer;

      // The C function expects uint8_t (*distribution_id)[16] - a pointer to a
      // 16-byte array. The binding incorrectly uses Pointer<Pointer<Uint8>>,
      // so we allocate a 16-byte buffer and cast it appropriately.
      final distIdData = calloc<Uint8>(16);
      distIdData.asTypedList(16).setAll(0, _distributionId);

      final msgPtr = calloc<Uint8>(plaintext.length);
      msgPtr.asTypedList(plaintext.length).setAll(0, plaintext);

      final buffer = calloc<SignalBorrowedBuffer>();
      buffer.ref.base = msgPtr.cast<UnsignedChar>();
      buffer.ref.length = plaintext.length;

      // Create the store struct with callbacks
      final storePtr = calloc<SignalSenderKeyStore>();
      storePtr.ref.ctx = Pointer.fromAddress(operationId);
      storePtr.ref.load_sender_key =
          Pointer.fromFunction<SignalLoadSenderKeyFunction>(
        _loadSenderKeyCallback,
        -1,
      );
      storePtr.ref.store_sender_key =
          Pointer.fromFunction<SignalStoreSenderKeyFunction>(
        _storeSenderKeyCallback,
        -1,
      );

      final storeConstPtr = calloc<SignalConstPointerFfiSenderKeyStoreStruct>();
      storeConstPtr.ref.raw = storePtr;

      try {
        final error = signal_group_encrypt_message(
          outPtr,
          senderPtr.ref,
          distIdData.cast<Pointer<Uint8>>(),
          buffer.ref,
          storeConstPtr.ref,
        );
        FfiHelpers.checkError(error, 'signal_group_encrypt_message');

        if (outPtr.ref.raw == nullptr) {
          throw LibSignalException.nullPointer('signal_group_encrypt_message');
        }

        // Save any updated sender key record
        final updatedRecordBytes = _pendingSenderKeyRecordBytes[operationId];
        if (updatedRecordBytes != null) {
          await _store.storeSenderKey(senderKeyName, updatedRecordBytes);
        }

        // Serialize the ciphertext message
        final resultPtr = calloc<SignalOwnedBuffer>();
        final ciphertextConstPtr = calloc<SignalConstPointerCiphertextMessage>();
        ciphertextConstPtr.ref.raw = outPtr.ref.raw;

        try {
          final serError = signal_ciphertext_message_serialize(
            resultPtr,
            ciphertextConstPtr.ref,
          );
          FfiHelpers.checkError(serError, 'signal_ciphertext_message_serialize');

          return FfiHelpers.fromOwnedBuffer(resultPtr.ref);
        } finally {
          calloc.free(resultPtr);
          calloc.free(ciphertextConstPtr);

          // Destroy the ciphertext message
          final destroyPtr = calloc<SignalMutPointerCiphertextMessage>();
          destroyPtr.ref.raw = outPtr.ref.raw;
          signal_ciphertext_message_destroy(destroyPtr.ref);
          calloc.free(destroyPtr);
        }
      } finally {
        // Securely zero plaintext before freeing
        LibSignalUtils.zeroBytes(msgPtr.asTypedList(plaintext.length));
        calloc.free(outPtr);
        calloc.free(senderPtr);
        calloc.free(distIdData);
        calloc.free(msgPtr);
        calloc.free(buffer);
        calloc.free(storePtr);
        calloc.free(storeConstPtr);
      }
    } finally {
      _pendingStoreOperations.remove(operationId);
      _pendingSenderKeyRecordBytes.remove(operationId);
    }
  }

  /// Decrypts a group message.
  ///
  /// The [senderAddress] is the address of the message sender.
  /// Returns the decrypted plaintext.
  Future<Uint8List> decrypt(
    ProtocolAddress senderAddress,
    Uint8List ciphertext,
  ) async {
    final operationId = _nextOperationId();

    // Pre-load existing sender key record
    final senderKeyName = SenderKeyName(
      senderAddress,
      _uuidToString(_distributionId),
    );
    final existingData = await _store.loadSenderKey(senderKeyName);
    if (existingData != null) {
      _pendingSenderKeyRecordBytes[operationId] = existingData;
    }

    _pendingStoreOperations[operationId] = _StoreOperation(
      _store,
      senderAddress,
      _distributionId,
    );

    try {
      final outPtr = calloc<SignalOwnedBuffer>();
      final senderPtr = calloc<SignalConstPointerProtocolAddress>();
      senderPtr.ref.raw = senderAddress.pointer;

      final msgPtr = calloc<Uint8>(ciphertext.length);
      msgPtr.asTypedList(ciphertext.length).setAll(0, ciphertext);

      final buffer = calloc<SignalBorrowedBuffer>();
      buffer.ref.base = msgPtr.cast<UnsignedChar>();
      buffer.ref.length = ciphertext.length;

      // Create the store struct with callbacks
      final storePtr = calloc<SignalSenderKeyStore>();
      storePtr.ref.ctx = Pointer.fromAddress(operationId);
      storePtr.ref.load_sender_key =
          Pointer.fromFunction<SignalLoadSenderKeyFunction>(
        _loadSenderKeyCallback,
        -1,
      );
      storePtr.ref.store_sender_key =
          Pointer.fromFunction<SignalStoreSenderKeyFunction>(
        _storeSenderKeyCallback,
        -1,
      );

      final storeConstPtr = calloc<SignalConstPointerFfiSenderKeyStoreStruct>();
      storeConstPtr.ref.raw = storePtr;

      try {
        final error = signal_group_decrypt_message(
          outPtr,
          senderPtr.ref,
          buffer.ref,
          storeConstPtr.ref,
        );
        FfiHelpers.checkError(error, 'signal_group_decrypt_message');

        // Save any updated sender key record
        final updatedRecordBytes = _pendingSenderKeyRecordBytes[operationId];
        if (updatedRecordBytes != null) {
          await _store.storeSenderKey(senderKeyName, updatedRecordBytes);
        }

        return FfiHelpers.fromOwnedBuffer(outPtr.ref);
      } finally {
        // Securely zero ciphertext before freeing
        LibSignalUtils.zeroBytes(msgPtr.asTypedList(ciphertext.length));
        calloc.free(outPtr);
        calloc.free(senderPtr);
        calloc.free(msgPtr);
        calloc.free(buffer);
        calloc.free(storePtr);
        calloc.free(storeConstPtr);
      }
    } finally {
      _pendingStoreOperations.remove(operationId);
      _pendingSenderKeyRecordBytes.remove(operationId);
    }
  }

  /// The sender address for this session.
  ProtocolAddress get senderAddress => _senderAddress;

  /// The distribution ID (UUID) for this group.
  Uint8List get distributionId => Uint8List.fromList(_distributionId);

  /// Converts a UUID string to bytes.
  static Uint8List uuidFromString(String uuid) => _stringToUuid(uuid);

  /// Converts UUID bytes to a string.
  static String uuidToString(Uint8List uuid) => _uuidToString(uuid);
}

File: rust/src/api/group_session.rs
//! Group session API for Signal Protocol group messaging (Sender Keys).
//!
//! This module provides group messaging operations using DartFn callbacks
//! for store operations. The hybrid approach:
//! 1. Receives Dart callbacks for loading/storing sender keys
//! 2. Uses in-memory stores temporarily during operations
//! 3. Returns results for Dart to persist
//!
//! # Security
//!
//! All sensitive cryptographic data is explicitly zeroed after use via the `zeroize` crate.

use flutter_rust_bridge::DartFnFuture;
use futures::executor::block_on;
use libsignal_protocol::{
    IdentityKeyPair, InMemSenderKeyStore, ProtocolAddress, SenderKeyDistributionMessage,
    SenderKeyRecord, SenderKeyStore,
    SignalProtocolError,
};
use rand::{TryRngCore as _, rngs::OsRng};
use uuid::Uuid;
use zeroize::Zeroize;

// ============================================================================
// GROUP KEY DISTRIBUTION with DartFn callbacks
// ============================================================================

/// Result of creating a sender key distribution message.
pub struct CreateSenderKeyDistributionResult {
    /// The sender key distribution message to share with group members.
    pub distribution_message: Vec<u8>,
    /// The updated sender key record to store.
    pub sender_key_record: Vec<u8>,
}

/// Create a new sender key distribution message with callbacks.
///
/// This creates or updates our sender key for a group and produces a
/// distribution message that should be sent to other group members.
///
/// # Callbacks
/// - `load_sender_key(sender_name, sender_device_id, distribution_id)` - Load existing sender key
/// - `store_sender_key(sender_name, sender_device_id, distribution_id, record)` - Store updated sender key
/// - `get_identity_key_pair()` - Get our identity key pair
///
/// # Returns
/// The distribution message to send to group members.
pub async fn create_sender_key_distribution_message_with_callbacks(
    sender_name: String,
    sender_device_id: u32,
    distribution_id: String,
    load_sender_key: impl Fn(String, u32, String) -> DartFnFuture<Option<Vec<u8>>> + Send + Sync + 'static,
    store_sender_key: impl Fn(String, u32, String, Vec<u8>) -> DartFnFuture<()> + Send + Sync + 'static,
    get_identity_key_pair: impl Fn() -> DartFnFuture<Vec<u8>> + Send + Sync + 'static,
) -> Result<CreateSenderKeyDistributionResult, String> {
    // Step 1: Load data via callbacks
    let mut existing_key_bytes = load_sender_key(sender_name.clone(), sender_device_id, distribution_id.clone()).await;
    let mut identity_key_pair_bytes = get_identity_key_pair().await;

    // Step 2: Call inner function
    let result = create_sender_key_distribution_inner(
        &sender_name,
        sender_device_id,
        &distribution_id,
        &existing_key_bytes,
        &identity_key_pair_bytes,
    );

    // SECURITY: Zeroize sensitive data
    identity_key_pair_bytes.zeroize();
    if let Some(ref mut bytes) = existing_key_bytes {
        bytes.zeroize();
    }

    let (distribution_message, sender_key_record) = result?;

    // Step 3: Store via callback
    store_sender_key(sender_name, sender_device_id, distribution_id, sender_key_record.clone()).await;

    Ok(CreateSenderKeyDistributionResult {
        distribution_message,
        sender_key_record,
    })
}

fn create_sender_key_distribution_inner(
    sender_name: &str,
    sender_device_id: u32,
    distribution_id: &str,
    existing_key_bytes: &Option<Vec<u8>>,
    identity_key_pair_bytes: &[u8],
) -> Result<(Vec<u8>, Vec<u8>), String> {
    // Parse identity (unused but validates the data)
    let _our_identity = IdentityKeyPair::try_from(identity_key_pair_bytes)
        .map_err(|e| e.to_string())?;

    // Parse distribution ID as UUID
    let dist_uuid = Uuid::parse_str(distribution_id)
        .map_err(|e| format!("Invalid distribution ID: {}", e))?;

    // Create protocol address
    let sender_address = ProtocolAddress::new(
        sender_name.to_string(),
        sender_device_id.try_into().map_err(|_| "Invalid device ID")?,
    );

    // Create in-memory store
    let mut sender_key_store = InMemSenderKeyStore::new();

    // Load existing sender key if present
    if let Some(bytes) = existing_key_bytes {
        let record = SenderKeyRecord::deserialize(bytes)
            .map_err(|e: SignalProtocolError| e.to_string())?;
        block_on(async {
            sender_key_store.store_sender_key(&sender_address, dist_uuid, &record).await
        }).map_err(|e| e.to_string())?;
    }

    // Create the distribution message
    let distribution_message = block_on(async {
        libsignal_protocol::create_sender_key_distribution_message(
            &sender_address,
            dist_uuid,
            &mut sender_key_store,
            &mut OsRng.unwrap_err(),
        ).await
    }).map_err(|e| e.to_string())?;

    // Get the updated sender key
    let updated_key = block_on(async {
        sender_key_store.load_sender_key(&sender_address, dist_uuid).await
    }).map_err(|e: SignalProtocolError| e.to_string())?
        .ok_or("Sender key not created")?;

    // Serialize results
    let distribution_bytes = distribution_message.serialized().to_vec();
    let key_bytes = updated_key.serialize().map_err(|e| e.to_string())?;

    Ok((distribution_bytes, key_bytes))
}

/// Process a sender key distribution message with callbacks.
///
/// This stores the sender's key so we can decrypt their messages.
///
/// # Callbacks
/// - `load_sender_key(sender_name, sender_device_id, distribution_id)` - Load existing sender key
/// - `store_sender_key(sender_name, sender_device_id, distribution_id, record)` - Store the sender key
///
/// # Returns
/// The sender key record to store.
pub async fn process_sender_key_distribution_message_with_callbacks(
    sender_name: String,
    sender_device_id: u32,
    distribution_id: String,
    distribution_message: Vec<u8>,
    load_sender_key: impl Fn(String, u32, String) -> DartFnFuture<Option<Vec<u8>>> + Send + Sync + 'static,
    store_sender_key: impl Fn(String, u32, String, Vec<u8>) -> DartFnFuture<()> + Send + Sync + 'static,
) -> Result<Vec<u8>, String> {
    // Step 1: Load existing key if any
    let mut existing_key_bytes = load_sender_key(sender_name.clone(), sender_device_id, distribution_id.clone()).await;

    // Step 2: Process
    let result = process_sender_key_distribution_inner(
        &sender_name,
        sender_device_id,
        &distribution_id,
        &distribution_message,
        &existing_key_bytes,
    );

    // SECURITY: Zeroize existing key
    if let Some(ref mut bytes) = existing_key_bytes {
        bytes.zeroize();
    }

    let sender_key_record = result?;

    // Step 3: Store via callback
    store_sender_key(sender_name, sender_device_id, distribution_id, sender_key_record.clone()).await;

    Ok(sender_key_record)
}

fn process_sender_key_distribution_inner(
    sender_name: &str,
    sender_device_id: u32,
    distribution_id: &str,
    distribution_message: &[u8],
    existing_key_bytes: &Option<Vec<u8>>,
) -> Result<Vec<u8>, String> {
    // Parse distribution ID as UUID
    let dist_uuid = Uuid::parse_str(distribution_id)
        .map_err(|e| format!("Invalid distribution ID: {}", e))?;

    // Create protocol address
    let sender_address = ProtocolAddress::new(
        sender_name.to_string(),
        sender_device_id.try_into().map_err(|_| "Invalid device ID")?,
    );

    // Parse the distribution message
    let skdm = SenderKeyDistributionMessage::try_from(distribution_message)
        .map_err(|e: SignalProtocolError| e.to_string())?;

    // Create in-memory store
    let mut sender_key_store = InMemSenderKeyStore::new();

    // Load existing sender key if present
    if let Some(bytes) = existing_key_bytes {
        let record = SenderKeyRecord::deserialize(bytes)
            .map_err(|e: SignalProtocolError| e.to_string())?;
        block_on(async {
            sender_key_store.store_sender_key(&sender_address, dist_uuid, &record).await
        }).map_err(|e| e.to_string())?;
    }

    // Process the distribution message
    block_on(async {
        libsignal_protocol::process_sender_key_distribution_message(
            &sender_address,
            &skdm,
            &mut sender_key_store,
        ).await
    }).map_err(|e| e.to_string())?;

    // Get the stored sender key
    let sender_key = block_on(async {
        sender_key_store.load_sender_key(&sender_address, dist_uuid).await
    }).map_err(|e: SignalProtocolError| e.to_string())?
        .ok_or("Sender key not stored after processing")?;

    sender_key.serialize().map_err(|e| e.to_string())
}

// ============================================================================
// GROUP ENCRYPTION / DECRYPTION with DartFn callbacks
// ============================================================================

/// Result of encrypting a group message.
pub struct GroupEncryptResult {
    /// The encrypted message.
    pub ciphertext: Vec<u8>,
    /// The updated sender key record.
    pub sender_key_record: Vec<u8>,
}

/// Encrypt a message to a group with callbacks.
///
/// # Callbacks
/// - `load_sender_key(sender_name, sender_device_id, distribution_id)` - Load our sender key
/// - `store_sender_key(sender_name, sender_device_id, distribution_id, record)` - Store updated sender key
/// - `get_identity_key_pair()` - Get our identity key pair (for signing)
///
/// # Returns
/// The encrypted ciphertext and updated sender key record.
pub async fn group_encrypt_with_callbacks(
    sender_name: String,
    sender_device_id: u32,
    distribution_id: String,
    plaintext: Vec<u8>,
    load_sender_key: impl Fn(String, u32, String) -> DartFnFuture<Option<Vec<u8>>> + Send + Sync + 'static,
    store_sender_key: impl Fn(String, u32, String, Vec<u8>) -> DartFnFuture<()> + Send + Sync + 'static,
    get_identity_key_pair: impl Fn() -> DartFnFuture<Vec<u8>> + Send + Sync + 'static,
) -> Result<GroupEncryptResult, String> {
    // Step 1: Load data
    let mut sender_key_bytes = load_sender_key(sender_name.clone(), sender_device_id, distribution_id.clone()).await;
    let mut identity_key_pair_bytes = get_identity_key_pair().await;

    // Step 2: Encrypt
    let result = group_encrypt_inner(
        &sender_name,
        sender_device_id,
        &distribution_id,
        &plaintext,
        &sender_key_bytes,
        &identity_key_pair_bytes,
    );

    // SECURITY: Zeroize sensitive data
    identity_key_pair_bytes.zeroize();
    if let Some(ref mut bytes) = sender_key_bytes {
        bytes.zeroize();
    }

    let (ciphertext, sender_key_record) = result?;

    // Step 3: Store updated key
    store_sender_key(sender_name, sender_device_id, distribution_id, sender_key_record.clone()).await;

    Ok(GroupEncryptResult {
        ciphertext,
        sender_key_record,
    })
}

fn group_encrypt_inner(
    sender_name: &str,
    sender_device_id: u32,
    distribution_id: &str,
    plaintext: &[u8],
    sender_key_bytes: &Option<Vec<u8>>,
    identity_key_pair_bytes: &[u8],
) -> Result<(Vec<u8>, Vec<u8>), String> {
    // Parse identity (unused but validates the data)
    let _our_identity = IdentityKeyPair::try_from(identity_key_pair_bytes)
        .map_err(|e| e.to_string())?;

    // Parse distribution ID as UUID
    let dist_uuid = Uuid::parse_str(distribution_id)
        .map_err(|e| format!("Invalid distribution ID: {}", e))?;

    // Create protocol address
    let sender_address = ProtocolAddress::new(
        sender_name.to_string(),
        sender_device_id.try_into().map_err(|_| "Invalid device ID")?,
    );

    // Create in-memory store
    let mut sender_key_store = InMemSenderKeyStore::new();

    // Load sender key
    let sender_key_bytes = sender_key_bytes.as_ref()
        .ok_or("No sender key found. Create a distribution message first.")?;
    let record = SenderKeyRecord::deserialize(sender_key_bytes)
        .map_err(|e: SignalProtocolError| e.to_string())?;
    block_on(async {
        sender_key_store.store_sender_key(&sender_address, dist_uuid, &record).await
    }).map_err(|e| e.to_string())?;

    // Encrypt
    let ciphertext = block_on(async {
        libsignal_protocol::group_encrypt(
            &mut sender_key_store,
            &sender_address,
            dist_uuid,
            plaintext,
            &mut OsRng.unwrap_err(),
        ).await
    }).map_err(|e| e.to_string())?;

    // Get updated sender key
    let updated_key = block_on(async {
        sender_key_store.load_sender_key(&sender_address, dist_uuid).await
    }).map_err(|e: SignalProtocolError| e.to_string())?
        .ok_or("Sender key not found after encryption")?;

    let key_bytes = updated_key.serialize().map_err(|e| e.to_string())?;

    Ok((ciphertext.serialized().to_vec(), key_bytes))
}

/// Result of decrypting a group message.
pub struct GroupDecryptResult {
    /// The decrypted plaintext.
    pub plaintext: Vec<u8>,
    /// The updated sender key record.
    pub sender_key_record: Vec<u8>,
}

/// Decrypt a message from a group member with callbacks.
///
/// # Callbacks
/// - `load_sender_key(sender_name, sender_device_id, distribution_id)` - Load the sender's key
/// - `store_sender_key(sender_name, sender_device_id, distribution_id, record)` - Store updated sender key
///
/// # Returns
/// The decrypted plaintext and updated sender key record.
pub async fn group_decrypt_with_callbacks(
    sender_name: String,
    sender_device_id: u32,
    distribution_id: String,
    ciphertext: Vec<u8>,
    load_sender_key: impl Fn(String, u32, String) -> DartFnFuture<Option<Vec<u8>>> + Send + Sync + 'static,
    store_sender_key: impl Fn(String, u32, String, Vec<u8>) -> DartFnFuture<()> + Send + Sync + 'static,
) -> Result<GroupDecryptResult, String> {
    // Step 1: Load sender key
    let mut sender_key_bytes = load_sender_key(sender_name.clone(), sender_device_id, distribution_id.clone()).await;

    // Step 2: Decrypt
    let result = group_decrypt_inner(
        &sender_name,
        sender_device_id,
        &distribution_id,
        &ciphertext,
        &sender_key_bytes,
    );

    // SECURITY: Zeroize loaded key
    if let Some(ref mut bytes) = sender_key_bytes {
        bytes.zeroize();
    }

    let (plaintext, sender_key_record) = result?;

    // Step 3: Store updated key
    store_sender_key(sender_name, sender_device_id, distribution_id, sender_key_record.clone()).await;

    Ok(GroupDecryptResult {
        plaintext,
        sender_key_record,
    })
}

fn group_decrypt_inner(
    sender_name: &str,
    sender_device_id: u32,
    distribution_id: &str,
    ciphertext: &[u8],
    sender_key_bytes: &Option<Vec<u8>>,
) -> Result<(Vec<u8>, Vec<u8>), String> {
    // Parse distribution ID as UUID
    let dist_uuid = Uuid::parse_str(distribution_id)
        .map_err(|e| format!("Invalid distribution ID: {}", e))?;

    // Create protocol address
    let sender_address = ProtocolAddress::new(
        sender_name.to_string(),
        sender_device_id.try_into().map_err(|_| "Invalid device ID")?,
    );

    // Create in-memory store
    let mut sender_key_store = InMemSenderKeyStore::new();

    // Load sender key
    let sender_key_bytes = sender_key_bytes.as_ref()
        .ok_or("No sender key found. Process a distribution message first.")?;
    let record = SenderKeyRecord::deserialize(sender_key_bytes)
        .map_err(|e: SignalProtocolError| e.to_string())?;
    block_on(async {
        sender_key_store.store_sender_key(&sender_address, dist_uuid, &record).await
    }).map_err(|e| e.to_string())?;

    // Decrypt
    let plaintext = block_on(async {
        libsignal_protocol::group_decrypt(
            ciphertext,
            &mut sender_key_store,
            &sender_address,
        ).await
    }).map_err(|e| e.to_string())?;

    // Get updated sender key
    let updated_key = block_on(async {
        sender_key_store.load_sender_key(&sender_address, dist_uuid).await
    }).map_err(|e: SignalProtocolError| e.to_string())?
        .ok_or("Sender key not found after decryption")?;

    let key_bytes = updated_key.serialize().map_err(|e| e.to_string())?;

    Ok((plaintext, key_bytes))
}
